### ISSUE TITLE
Blocking on approval

### ISSUE THREAD

**Issue by iHiD** [exercism/exercism#4908]
I asked a few experienced mentors the following question:

> Question: With the hindsight of the last year, do you think it's good that we block progress until a mentor approves a solution? Pros/cons?

These are the responses

---

Pros:

- A (far) better learning experience. Being blocked on the core exercises means that you _must_ learn the concepts the track wants you to learn before continuing. This prevents someone from just rushing through the exercises with their partial knowledge of the language, meaning that they could be doing things wrong/non-idiomatic in many exercises, instead of only one. Think of it like doing 1 or 3 week sprints. If you will get feedback after 1 week, there is far less chance of you having deviated much from the optimal track, whereas getting feedback after 3 weeks could result in you having to do more backpedalling.
- I feel that an exercise being “approved” and being able to move on the next exercise, is a great feeling of progression. It is reminiscent of completing a level in a video game. You are progressing and you know you are becoming more skilled.
- I think it is likely that people are spending more time with the core exercises, than they would have if they wouldn’t be blocked. There is more of an incentive to keep working on an exercise if you know you will be blocked from continuing if it is not okay.

Cons:

- The most obvious one is that one can be blocked from progressing for a (really) long time if a track is not being actively mentored. This can be quite frustrating at times, especially when getting stuck on one of the earlier exercises, as you usually won’t have unlocked many side exercises already.
- I don’t personally feel like this, but some people might feel that they are being “babysitted.” Why not just let me continue? I think the independent mode solved this though.
- Maybe it puts a bit more pressure on mentors, knowing that they _will_ have to look at a solution as the student cannot otherwise continue.
All in all, I was wholeheartedly in favor of doing this when we started with v2. However, I have also experienced the downsides of this system, having to wait for 5-7 days (sometimes weeks) before being able to continue, is a sub-optimal experience. It was also consistently the only negative feedback I got from people I introduced to exercism: the waiting time can be quite long (although this also did include side exercises). I feel this is only a problem for those tracks that have slow response times, for the other tracks I was perfectly fine with waiting a day or two being being allowed to continue.

---

Pros:
- mentoring has more meaning. If we block, students are waiting. So as a mentor I feel more obliged to do something.
- as a student I feel am more dependent on approval and therefore will work more (and learn more).
- The queues balance themselves. Yes, students jump ship or they switch to independent mode but better than all students waiting ages because queues don’t balance themselves any more.

Cons:
- students have to wait and might jump ship. But they can still use unmentored mode. Maybe point that out more direct while waiting.
- it is not so scalable. There are too many students for the mentors. I don’t feel like not blocking students is the answer, though. It might lead to students getting feedback any more as they just continue with the exercises and the queues overflow. Also we are trying to get it scalable.

So… that is my reasoning for why I think blocking the students is a central part of exercism. I’m not sure what it would be without it. I just know I like it a lot more than the first version where there was no real point in commenting on other students solutions. I tried but there never was anything like real learning happening.

---

I teach programming for a living, and I find that, yes, stopping them from automatically going, means that you don't have to address the same "nitpicky" kind of things in later exercises  After the 3rd week, I don't "stop" them from completing "future work" as much, unless they still haven't formed strong opinions about certain things.  I don't even really care which side of things they go with, as long as they have formed some kind of opinion on them.  (This doesn't happen very often that they haven't, when I have a dedicated set of students, but this is a little harder when you aren't with them necessarily throughout the journey, such as happens with Exercism.)  Early on,  I emulate Exercism's idea of mentoring and not letting them continue on, (but with Exercism I will approve early (sacrificing a rating, probably)), and let them continue submitting as we talk through exercises, and allow them to move to something a little harder (knowing I won't have to talk to them about some of the "easier and optional" things.

---

bads:
- student can not move on if mentor itself stuck.
being unable to proceed because of simple task is not cool.
that can completely discourage student.
Not each task is interconnected with previous tasks,
so the knowledge of previous tasks is important
but not something the progress should be stop because of.
- student can't get help if he is on queue but there are no mentors,
like some other tracks whith low numbet of mentors.

goods:
- student will not make a lot of mistakes. The process is kind of
sandboxed, mentor will always help †
- student knows on what to work, gradualy, step by step.
it helps focus.

† Actually I'm not sure if that it is good.
For example the book "Learn C the Hard Way" by Zed Shaw:
In that book strong emphasis put on the errors.
The pass of trials and errors makes knowledge deeply
rooted into the memory. So I'm not quite sure about that "sandboxins"
as a good feature in long term, but definitely good in short one.

---

For myself, using Exercism over the last year I have had some really great chats with mentors that lead me to start writing better code and learning more about the language. Sadly, most of that time that doesn't happen, I find myself blocked waiting, or not getting much feedback on the language vs how the mentor would solve the exercise. Does become a bit frustrating.

---

I dont know about wether or not I think its good. I think pros and cons are relatively clear. Pro is you keep the queue manageable (to an extent) and prevent students from moving on to make mistakes that they might ultimately feel that they might not have made if they had the opportunity to get feedback on a solution they had already submitted.

Con is that a student cant do anything and is left waiting on a mentor to make forward progress. But that making further progress might make the queue of pending solutions bloat with 10 or more solutions from one student
I suppose my hot take is ¯\_(ツ)_/¯

---

First impulse is *yes* -- it is good that we block progress.  *I didn't like it at first* - it seemed designed to slow things down and it frustrated me (as a student) - especially early on in the track.  Also?  What do these random people really know about me and my code??  _*I'm in a hurry to prove I've got this*_  and how do I know what a mentor says is any good?  Also - how do I know if they're even going to give me a review?  How long does this process take?

As a mentor, it's very disheartening to see the queue get larger and larger - no matter how many solutions I mentor/don't mentor.  It can also be frustrating to deal with students who quite clearly only want positive feedback (or think that passing the tests by any means necessary is the goal), and don't want to adapt/modify their code based on a reviewers perspective.  I also sometimes feel a little bit lost as to how far I should go with things - do I poke students on making tests of their own?  Do I recommend they attempt to log things?  I worry that I don't push the more advanced students far enough ....and I worry that I might scare off or confuse programming beginners.  The difference between how many solutions I've given feedback on vs how many I've "helped" (e.g. how many students have completed and moved on) is much larger than I'd like...how do I stop that from happening?
*BUT*
I was missing the point - in fact, I think _most people_ miss or forget the point.  This is less about coding up the solutions and more about what happens _after_ you get the tests to pass/after you get stuck.  Mentorship is about learning to have a productive conversation around your code/problem solving and other programmers code/problem solving.  About setting up a low-threat situation where you can really dig into the details of things and seriously talk about learning and improving.  There are no shortage of sites where you can solve problems and test your code...and there are no shortage of tutorials and projects out on the web.  But there aren't many (unpaid) places for getting honest feedback from another human that's not your professor, your friend, or your boss - so they aren't carrying around those particular biases, one way or another.  And it's real-world :  the mentor is bound to have a different perspective and coding style than you, so in a sense you are negotiating what the correct solution is going to be.  There's learning the syntax of a language...and then there's learning how programmers in that language see problems - what's common, what's idiomatic, what the language is "good for".  Other programmers might not tell you that - and you might not have the time to dig in if you are under the gun with something at work.  But you can do that here.  More than that - you can also learn and practice how to *give* feedback in a way that get's your point across, and convinces/motivates a student to change things productively.
Aaaall that said - the site might not be clear enough on a student's options.  Maybe move toward being very up front about what being on the mentored track means timing-wise and process wise?  Let students know they can just blow through things un-mentored, but that they are missing a great learning opportunity if they do so.  Allow students to skip a certain number of mentored solutions if they need to (but if they skip too many, maybe message them that they appear to be opting out of mentorship??)....but these thoughts are muddier for me...I need to sit a bit and look at the site, and maybe write up a more formal pro/con list.

I had an experience last week on codebuddies (another Slack community I am involved in) where a student in the Python channel asked for help working with "matrixes".....turns out, they were working on the `Matrix` problem for Exercism.  After walking through the solution that's on GitHub (wasn't my intent to point them to that...I thought they were working on something else at the time), the person said "it's too bad that Exercism doesn't have a systems design or OOP track ...like maybe for advanced users".  And I told them that that stuff usually comes out in the course of going through the main language tracks and getting feedback from mentors....but it was really interesting, because clearly they didn't see the site as having humans that would help them with that sort of thing, and they also saw the "learning" as only passing the tests and fulfilling the stated problem brief...a series of checklists that they marched through.  Dunno where I am headed with that, quite...but there is something there messaging-wise.

---

I do, yes :) It can be frustrating for both mentor and student sometimes, but I think it's essential that mentors have *some* leverage to get students to make changes. Otherwise I think some students will just submit crap solutions, ignore the comments, go on to the next exercise, submit a crap solution, and so on. They'll graduate the Exercism track not really having learned anything, but instead having just had a little code practice. Well, there are plenty of code practice sites; what makes Exercism different is the human feedback.

---

Pros:
- Misunderstandings about Exercism workflow (submitting the wrong file, confusion on how to link a solution against the tests, etc) or a certain language feature are caught early, so the student doesn't have to retrace through every exercise they did incorrectly by the time they receive help.
- As the focus of Exercism is language fluency, it is important to learn concepts in an idiomatic way, not necessarily just the first way that "works". A student might get through several exercises, running tests correctly, but using some methodology that is fundamentally backwards to how that language works. When you stack this with the current length of the mentor queues (which would become more full of "trivially bad" submissions since confused students wouldn't be blocked), these students might even "complete" the track before they get the help they need, and move on, never checking back in for the mentor feedback. This produces students who "learned through Exercism" that may not really know what they're doing; it hurts the Exercism brand, sure, but more importantly, these students might get a false confidence that leads them into unfortunate situations later.

Cons:
- Mentor queues are long. Some students will abandon Exercism when they're tired of being blocked.
- Some students don't take "good" criticism well, and will quit instead of working *with* the mentor through multiple submission iterations
- Occasionally, a mentor might come across too harsh (or be perceived that way), turning the student off from the platform altogether

Overall, I'd say blocking progress is the right way to go, but I understand the frustration from students who just want to blaze through the exercises uninhibited; independent mode is probably best for some of these, but if you don't know the mode already exists, you might not scroll down the page to see the button to switch; perhaps a quick one-time dialog could pop up when starting a track, asking the student to choose Mentored or Independent? There could be a message on the benefits of each, and a note that those already experienced in the language might prefer the freedom of independent mode.

---

Hmm. I think it depends almost entirely on queue length (in terms of time). Being blocked certainly makes you more invested. But it also means, especially with mentoring notes, that students are pushed towards one singular ideal solution. that means you don't have the diversity in community solutions, for better or worse, that you used to. I think it would be really interesting to offer some students unblocking mentorship and see how that affects student and mentor satisfaction.

---

I think there’s pros and cons for each - pro for blocking includes that the student’s overall quality *should* be improving as the move through the course. They will presumably have received feedback on what they should be doing each time and hopefully each exercise adds a little bit more that they should be considering and doing when submitting. If they are not blocked, then they could submit most/all solutions and have the same problems that should have been caught and called out earlier on. On the other hand allowing students to move on without blocking means they can go at their own pace which could increase their ability to learn the language. If feedback is quick (probably quicker than we can reasonably expect with human mentors in almost all cases) then it is conducive to continuing right away and learning. When there’s days or weeks or more between a submission and feedback, it can be like starting over on learning a language unless the student is also doing more than just exercism for their learning. I’ve personally had a bit of frustration as a student where it seems that no matter what I’m doing there’s at least one more thing that the mentor wants and could have mentioned with the first submission. Combine that with a few days between submission and feedback and students can be “stuck” on an exercise for quite some time.

---

Very good question, glad that you followed up on this. I still think that blocking progress is not desirable — that is it would be nice if users could work on future exercises in the track without being blocked on submission.

For me it boils down to two key factors: momentum and linear learning path
By momentum, I mean the ability to keep users engaged in exercism over time. I suspect that many of our users are not full-time students, instead they have regular jobs and find time to do exercism here and there — on the evenings, on the weekends, etc. When you have to wait a long period of time after you submit an exercise, it’s easy to simply move on to some other activity and not come back to it. That’s not a great way to learn, and not a great way to stay engaged with a website.
Realistically with mentors being volunteers, the turnaround time can be long — from several hours to several days (or even longer) and that really kills the momentum that someone has when using the site. I’ve seen this a few times as a mentor — I’ll post some comments on a solution and then never hear from the student again — we’ve lost that student, possibly forever.
I’ve also heard this anecdotally from other people I know who’ve used exercism — someone I know tried the golang track but switched to mentored mode so he could go through a lot of exercises all at once. It’s just slow and frustrating to do an exercise and then have to wait for days for a response but then have nothing else to do, especially in a world where a lot of things happen in real time.
Yes, I know there are those optional exercises, but those feel like fillers and it feels weird to be able to do as many of those as you want but not do as many core exercises as you want.

on the linear learning point — I find it a little bit frustrating as a student to have to go down a linear track. I understand that people like maud have put a lot of thought into making the tracks build on each other, but as a student I like being able to choose my own way of learning things — and being forced down a linear path feels like sort of an antiquated way of learning
I imagine many other students are the same
when going through online learning courses like on Codecademy, the learning tracks allow you to select different courses along the track — some courses have prerequisites (I think) but you can choose to study different topics within a track.
I am not an educator but it seems to me that non-linear learning is a more effective way of learning than a traditional linear curriculum

---

yes and no, if you feel a student should try out a piece of code to understand it so you can give further feedback then yes. If it is syntax/naming/convention then we should inform and let them continue unless they feel they want further feedback.

---

I’m sorry if “it depends” is not the answer you are looking for but…

Pro Blocking:

•    Blocking provides a clear discipline and structure for progress through the track.  The mentee has some insight into how well they are doing.

•    Blocking (at least the unblocking part) provides a sense of achievement, a very tangible reward for tasks accomplished.

Con Blocking:

•    The speed of processing queues is not consistently adequate so mentees’ progress is often stalled.

•    Philosophically mentoring is / should be / will be much more than reviewing exercises.  “Approval” already seems the wrong word (not that I’m suggesting something so easily understandable should be changed).  As a mentor I tend to disapprove very little (using some well-defined criteria).  Not blocking would be more consistent with the idea of supporting mentees rather than passing judgement on them.

On Balance:

•    I would review the processing speed of the tracks over the last year and make a call based on that.  Is mentee retention on well performing tracks better than poorly performing ones?  Presumably the same insights can be gleaned from times when tracks perform well against those when they performed poorly.  If there is a significant difference you could hypothesise that speed of unblocking is a contributory factor and modify the unblocking rules.

Possible Alternatives (if you can stand the added complexity):

•    Unlock side exercises on a timed basis if a solution not picked up by a mentor in a timely fashion.

•    Unlock side exercises when sufficient preceding side exercises have been completed.

---

I'm still with you, so it seems to be working out much better than I expected :wink: In my opinion the mentoring aspect is what makes us "better" than the other "solve the puzzle" sites out there. I have not yet seen any other where discussions may evolve around a solution. The only thing I'm missing in the new exercism were the random discussions on code one found randomly. But to be honest, due a shortage in everything (especially time) I have not yet looked into community solutions since discussions have been enabled.

---

'll assume they've (the student) has come here for the novel feature we offer, guidance from a real person (even auto mentoring is still personal).  That isn't easy to come by for "$0".  They must be somewhat serious in their desire to learn and receive feedback or else why not go to Project Euler or Cyber-dojo etc if only interested in solving problems.  My initial thought to the change was that people won't like being held up.  What I finally came to realize is holding up a student does force that student and teacher interaction to take place that the student desires.  With the old system that could happen but I think it didn't more often than it did.  It would be nice if the student is in it to learn that they be open to shifting to other tracks, where wait times are shorter.  In some cases what they learn in other tracks is probably very applicable to their track of choice, that everyone else and their uncle is trying to do at the same time.  Perhaps it would be a good feature if students could see wait times for all tracks, as a way to try encourage some diversity in thought.  Kinda like when pulling up to toll station, who seeks out the longest line; not I.  

---

Generally speaking I think it's a good change because that kind of feedback makes Exercism unique and #testimonials makes me think that it's one of the main reasons the platform grew that much.

Being "forced" to listen to feedback and improve your solution strenghtens the learning process, imo. In particular if the mentor guides you towards the solution and discusses what to improve with you, but you still come up with it yourself after some "struggling". This is a pretty effective learning technique in general. I think that if waiting before progressing was voluntary, many people would not benefit from it because it's very easy to trick yourself into thinking that you understood certain feedback and will do better next time (from personal experience, unless you actually apply the feedback right then, you likely won't do it in the future either).

However, in particular on smaller tracks, e.g. Julia, there's a lot of pressure and burden on the mentor(s). If there aren't enough mentors, it might be better to (temporarily) switch back to the v1 system of being able to optionally request feedback. Relying on 1-2 mentors is not sustainable at all. However, Julia and its community is somewhat exceptional in many ways[1] and I wouldn't be surprised if it's a bit easier for other languages to find mentors. In that case, what I've described is likely not a huge problem.
While waiting time is a big issue, the track anatomy project and auto mentoring will likely shorten it a lot. I still think it can be a problem if the wait is too long and it has certainly driven some people away from the platform but it seems solvable.

So TL;DR: I think it's good that progress is blocked, but it might need some fine-tuning on certain tracks.

[Disclaimer: I've only mentored 1 solution or so in the last few months because I'm unable to commit so much time "in advance", so this is purely based on discussions, observing the track and such.]

[1] A huge part of the community is in academia (there's a running joke that procrastinating PhD students are the main reason the ecosystem is thriving :smile:). That also means that many Julia users teach as part of their job and give workshops etc. so the willingness to become a mentor in their free-time is rather low (for understandable reasons!).

---

Pro: Approving at the right time increases motivation to improve a solution.

Con: But it also increases the risk of tension if the mentor thinks something should be improved that the student doesn't care about.

I started the Erlang track anonymously and on the 'Bob' exercise was asked to use function guards instead of 'if' because function guards are generally more useful and idiomatic in Erlang. ('If' isn't really an 'if' but more like a switch where each case is a guard/body rather than pattern/body.) But for the particular exercise, either one was passable. Had I mentored it, I'd have approved it but suggested to use function guards (and had it been made with function guards, I'd have approved it and suggested 'if'). So when it wasn't approved, I felt provoked, because I've been teaching Erlang for 5 years, so I submitted all variations of reasonable solutions I could think of and asked if he preferred I did that when I don't know which one he will pass. Someone else might have given a lower score.

My point is: Finding out when to pass a solution and ask for more, and when to require more, is delicate, track-specific, exercise-specific, ties into the way you mentor, and is hard to quantify. So if we should have it, we have to socialize / make visible the way we mentor.

E.g. statistics visible to other mentors on what kind of mentoring behavior that triggers an adequate solution and a 5-star rating.

----

Pros of blocking include enforcing the track anatomy and key concepts that are learned along the way. Cons of blocking include getting stuck waiting for mentoring which seems to happen often based on the github issues.

I think blocking also helps maintain the size of the mentor queue, which would explode in size if students weren’t blocked by mentoring.

Currently we have a good solution I think. Independent mode allows students to request mentoring, but to also continue in the track

---

I think if we structure it more like a classroom, allow students to do as many exercises as they can, then the mentors can provide feedback when possible without breaking the flow of learning.

The other consideration is the tests are the most immediate feedback the students get on whether or not they are going down a correct path.

So while the tests pass, the mentors can give more insight or advice on the using the language to solve problems.

I think for tracks with less mentors that'll help keep students active. I think for the tracks with a lot of active mentors, that approach is fine and could be preferential to the students, but can be a turnoff in the tracks with less mentors.

---

I’d tend to say yes, but with the caveat that that’s really _only_ true if the queue is short and the time delay to first contact with a mentor on any given exercise is <24 hours, ideally <8.

Average wait time of the 20 top exercises right now on Python is 28 days. If that holds true for all 17 core exercises, it takes a student 1.3 years to make it through the Core. Though I can’t speak for other mentors I’ll move an engaged student who iterates rapidly through an average of 3-4 iterations to approval, plus 2-3 iterations of post-approval refinement. It’s a guess, but I’d think most of my students get from first contact to approval with waiting on me for a maximum of 24 hours, and that’s when I’m moving the boat and physically can’t use the machine. That means they’re progress blocked awaiting mentor approval for 17 days while they’re waiting on queue churn for ~450 days.

If they could progress through core in under a month then blocking progress is an obvious good, just as a check on Dunning-Kruger and a chance to experience the benefits of Mentoring to their fluency. If an actively engaged student with a serious goal whose set aside some time to get up to speed _can’t_ make it though the 17 core exercises in less than a month, then blocking progress is an insurmountable disincentive.

At the same time I wonder if there isn’t something to be said for instituting some sort of backoff policy. First two gateway exercises block all the way through to approval, but the following ones unlock once you’ve been in the queue for some amount of time, doing so after progressively shorter timeouts the higher you are up the core progression. That way the shorter the queues get the more students will have to wait till they’ve gotten meaningful feedback to proceed, while with longer queues students can progress but go back to previous still un-approved exercises as mentor eyes become available.

---

Sorry for taking a bit to respond - still in Amsterdam in tourist mode. I think that with the Track Anatomy work it’s good that we block progression. From my understanding much of Maud’s work is around the idea that students are practicing certain things before moving to the next exercise and more often than not we’d want them to be mentored on the simpler things before moving on to more complex things. In the extreme a student with low fluency could go through the entire track very quickly and mentors would need to re-address the same issues over an over because they were never thought otherwise in the earlier exercises.

---

*Cons*:

- Optimal solutions are blocked and _have to wait_. This will partially be solved by Automated Mentoring Support (AMS), and partially ís solved by side exercises.
- Mentoring pressure. Mentoring is _mandatory_ so that means that the queue is actually people being *blocked* (if you forget side exercises).
- Long wait time can be de-captivating and de-motivating. AMS partially solves this.
- People who know what they're doing can get frustrated, but that's what independent mode is for.

*Pros*:

- In lieu with Track Anatomy (TA), there are certain subject we _want_ to be taught which we can't _always_ convey via AMS. Curently AMS is not completely there yet. Approving people automagically (as in, approve to continue) means that they don't necessarily pick up on the lesson. This can cause a bad waterfall e/affect where they don't apply the rules meant to be taught down the line, _straining the mentors_.
- "Visuals" in the sense that it makes it feel like you're _actually_ getting real mentorship. Blocking you makes this even more enforced and thus has a higher impact on feeling.
- Gatekeeping and flow control. By teaching student to logarithmically or half-baysianly solve the queue from bottom to top, doing it this way, we can keep students out of the pipe if it's too busy. Opening it up (auto approve and then wait) just means MORE people who have to wait > 200 days.

---
Pros:
- mentoring has more meaning. If we block, students are waiting. So as a mentor I feel more obliged to do something.
- as a student I feel am more dependent on approval and therefore will work more (and learn more).
- The queues balance themselves. Yes, students jump ship or they switch to independent mode but better than all students waiting ages because queues don’t balance themselves any more.

Cons:
- students have to wait and might jump ship. But they can still use unmentored mode. Maybe point that out more direct while waiting.
- it is not so scalable. There are too many students for the mentors. I don’t feel like not blocking students is the answer, though. It might lead to students getting feedback any more as they just continue with the exercises and the queues overflow. Also we are trying to get it scalable.

So… that is my reasoning for why I think blocking the students is a central part of exercism. I’m not sure what it would be without it. I just know I like it a lot more than the first version where there was no real point in commenting on other students solutions. I tried but there never was anything like real learning happening.

**Comment by ErikSchierboom** [comment]
Hello 👋

With the launch of Exercism v3, we are **closing all issues in this repository** to help give us a clean slate to detect new problems. If this issue is still relevant to Exercism v3 (e.g. it's a feature that we haven't implemented in v3, or a bug that still exists), please reopen it and we will review it and post an update on it as soon as we get chance.

Thanks for helping make Exercism better, and we hope you enjoy v3 🙂
